### ISSUE TITLE
machine-config-daemon firstboot-complete-machineconfig generate segmentation violation code=0x1

### ISSUE THREAD

**Issue by olafWW** [openshift/machine-config-operator#2696]
OCP cluster initial deployment does not success... It get stuck on the master nodes when running:
machine-config-daemon-firstboot

**Description**

**Steps to reproduce the issue:**
1. bare metal deployment of an openshift cluster 
2. login on the master node , once it is reachable 
3. monitor whats going on:

**Describe the results you received:**
[root@master1 ~]# systemctl status machine-config-daemon-firstboot
● machine-config-daemon-firstboot.service - Machine Config Daemon Firstboot
   Loaded: loaded (/etc/systemd/system/machine-config-daemon-firstboot.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2021-07-29 10:23:29 UTC; 7min ago
  Process: 2048 ExecStart=/run/bin/machine-config-daemon firstboot-complete-machineconfig (code=exited, status=2)
 Main PID: 2048 (code=exited, status=2)
      CPU: 52.662s

Jul 29 10:23:29 master1.ocp4.beer4laff.de machine-config-daemon[2048]: github.com/spf13/cobra.(*Command).ExecuteC(0x2b066c0, 0xc000737f78, 0x4062c5, 0xc00007e058)
Jul 29 10:23:29 master1.ocp4.beer4laff.de machine-config-daemon[2048]:         /go/src/github.com/openshift/machine-config-operator/vendor/github.com/spf13/cobra/command.go:958 +0x375
Jul 29 10:23:29 master1.ocp4.beer4laff.de machine-config-daemon[2048]: github.com/spf13/cobra.(*Command).Execute(...)
Jul 29 10:23:29 master1.ocp4.beer4laff.de machine-config-daemon[2048]:         /go/src/github.com/openshift/machine-config-operator/vendor/github.com/spf13/cobra/command.go:895
Jul 29 10:23:29 master1.ocp4.beer4laff.de machine-config-daemon[2048]: main.main()
Jul 29 10:23:29 master1.ocp4.beer4laff.de machine-config-daemon[2048]:         /go/src/github.com/openshift/machine-config-operator/cmd/machine-config-daemon/main.go:27 +0x31
Jul 29 10:23:29 master1.ocp4.beer4laff.de systemd[1]: machine-config-daemon-firstboot.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jul 29 10:23:29 master1.ocp4.beer4laff.de systemd[1]: machine-config-daemon-firstboot.service: Failed with result 'exit-code'.
Jul 29 10:23:29 master1.ocp4.beer4laff.de systemd[1]: Failed to start Machine Config Daemon Firstboot.
Jul 29 10:23:29 master1.ocp4.beer4laff.de systemd[1]: machine-config-daemon-firstboot.service: Consumed 52.662s CPU time
[root@master1 ~]# 

running the command manually  unveils:

[root@master1 ~]# /run/bin/machine-config-daemon firstboot-complete-machineconfig
I0729 11:09:45.459684    2498 rpm-ostree.go:261] Running captured: rpm-ostree status --json
I0729 11:09:45.527671    2498 daemon.go:218] Booted osImageURL:  (47.83.202105220305-0)
I0729 11:09:45.529000    2498 update.go:630] Checking Reconcilable for config mco-empty-mc to rendered-master-085d53fdda367a73d6034ee3723ab891
I0729 11:09:45.529851    2498 update.go:1943] Starting update from mco-empty-mc to rendered-master-085d53fdda367a73d6034ee3723ab891: &{osUpdate:true kargs:false fips:false passwd:false files:false units:false kernelType:false extensions:false}
I0729 11:09:45.532525    2498 update.go:1258] Updating files
I0729 11:09:45.532782    2498 update.go:1331] Deleting stale data
I0729 11:09:45.534072    2498 run.go:18] Running: nice -- ionice -c 3 oc image extract --path /:/run/mco-machine-os-content/os-content-746187922 --registry-config /var/lib/kubelet/config.json quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:07c0cbfed7fb3efdf794535430971f8d78fca8fde8f040ed38bdc5b435075044

I0729 11:12:56.934450    2498 update.go:1821] Updating OS to quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:07c0cbfed7fb3efdf794535430971f8d78fca8fde8f040ed38bdc5b435075044
I0729 11:12:56.934476    2498 rpm-ostree.go:261] Running captured: rpm-ostree status --json
I0729 11:12:56.989449    2498 rpm-ostree.go:184] Current origin is not custom
I0729 11:12:58.845167    2498 rpm-ostree.go:211] Pivoting to: 47.83.202106252242-0 (f228e5682bf91e1b70394d17e57c896d8966a5841bb850806c29060a253eea44)
I0729 11:12:58.845185    2498 rpm-ostree.go:243] Executing rebase from repo path /run/mco-machine-os-content/os-content-746187922/srv/repo with customImageURL pivot://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:07c0cbfed7fb3efdf794535430971f8d78fca8fde8f040ed38bdc5b435075044 and checksum f228e5682bf91e1b70394d17e57c896d8966a5841bb850806c29060a253eea44
I0729 11:12:58.845194    2498 rpm-ostree.go:261] Running captured: rpm-ostree rebase --experimental /run/mco-machine-os-content/os-content-746187922/srv/repo:f228e5682bf91e1b70394d17e57c896d8966a5841bb850806c29060a253eea44 --custom-origin-url pivot://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:07c0cbfed7fb3efdf794535430971f8d78fca8fde8f040ed38bdc5b435075044 --custom-origin-description Managed by machine-config-operator
**panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1895ca1]**

goroutine 1 [running]:
github.com/openshift/machine-config-operator/pkg/daemon.(*Daemon).applyOSChanges(0xc0001c22c0, 0xc000266ea0, 0xc0004b29c0, 0x0, 0x0)
        /go/src/github.com/openshift/machine-config-operator/pkg/daemon/update.go:465 +0x2c1
github.com/openshift/machine-config-operator/pkg/daemon.(*Daemon).update(0xc0001c22c0, 0xc000266ea0, 0xc0004b29c0, 0x0, 0x0)
        /go/src/github.com/openshift/machine-config-operator/pkg/daemon/update.go:712 +0xb58
github.com/openshift/machine-config-operator/pkg/daemon.(*Daemon).RunFirstbootCompleteMachineconfig(0xc0001c22c0, 0x2b51c88, 0xc000058060)
        /go/src/github.com/openshift/machine-config-operator/pkg/daemon/daemon.go:555 +0x173
main.runFirstBootCompleteMachineConfig(0x2b06420, 0x2b51c88, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/openshift/machine-config-operator/cmd/machine-config-daemon/firstboot_complete_machineconfig.go:39 +0x16f
main.executeFirstbootCompleteMachineConfig(0x2b06420, 0x2b51c88, 0x0, 0x0)
        /go/src/github.com/openshift/machine-config-operator/cmd/machine-config-daemon/firstboot_complete_machineconfig.go:43 +0x53
github.com/spf13/cobra.(*Command).execute(0x2b06420, 0x2b51c88, 0x0, 0x0, 0x2b06420, 0x2b51c88)
        /go/src/github.com/openshift/machine-config-operator/vendor/github.com/spf13/cobra/command.go:854 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x2b066c0, 0xc000717f78, 0x4062c5, 0xc000114058)
        /go/src/github.com/openshift/machine-config-operator/vendor/github.com/spf13/cobra/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /go/src/github.com/openshift/machine-config-operator/vendor/github.com/spf13/cobra/command.go:895
main.main()
        /go/src/github.com/openshift/machine-config-operator/cmd/machine-config-daemon/main.go:27 +0x31
[root@master1 ~]# 

**Describe the results you expected:**
I would expect, that the OCP deployment goes through, but it seems it gets stuck because of this failing service on the master nodes..

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `oc adm release info --commits | grep machine-config-operator`:**

```

[root@r84n1 .kube]# oc adm release info --commits | grep machine-config-operator
  machine-config-operator                        https://github.com/openshift/machine-config-operator                        8530c27d3d9b6155923d348058bc025a6a98ec3c
[root@r84n1 .kube]# 
```

**Additional environment details (platform, options, etc.):**
x86_64, 16 GB RAM, 4 COREs ...

**Comment by apoorvajagtap** [comment]
Hello, As per the goroutine dump and errors, it seems that some SEG fault is encountered while:
~~~
rpm-ostree rebase --experimental /run/mco-machine-os-content/os-content-746187922/srv/repo:f228e5682bf91e1b70394d17e57c896d8966a5841bb850806c29060a253eea44 --custom-origin-url pivot://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:07c0cbfed7fb3efdf794535430971f8d78fca8fde8f040ed38bdc5b435075044 --custom-origin-description Managed by machine-config-operator
~~~

- Can you please share the logs of rpm-ostreed.service from the respective node?
$ journalctl -xu rpm-ostreed.service

**Comment by olafWW** [comment]
while first-boot is still  running:

[core@master1 ~]$ journalctl -xu rpm-ostreed.service
-- Logs begin at Fri 2021-07-30 06:02:51 UTC, end at Fri 2021-07-30 06:11:42 UTC. --
Jul 30 06:10:38 master1.ocp4.beer4laff.de systemd[1]: Starting rpm-ostree System Management Daemon...
-- Subject: Unit rpm-ostreed.service has begun start-up
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
-- 
-- Unit rpm-ostreed.service has begun starting up.
Jul 30 06:10:38 master1.ocp4.beer4laff.de rpm-ostree[2037]: Reading config file '/etc/rpm-ostreed.conf'
Jul 30 06:10:38 master1.ocp4.beer4laff.de rpm-ostree[2037]: In idle state; will auto-exit in 62 seconds
Jul 30 06:10:38 master1.ocp4.beer4laff.de systemd[1]: Started rpm-ostree System Management Daemon.
-- Subject: Unit rpm-ostreed.service has finished start-up
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
-- 
-- Unit rpm-ostreed.service has finished starting up.
-- 
-- The start-up result is done.
Jul 30 06:10:38 master1.ocp4.beer4laff.de rpm-ostree[2037]: client(id:cli dbus:1.55 unit:machine-config-daemon-firstboot.service uid:0) added; new total=1
Jul 30 06:10:38 master1.ocp4.beer4laff.de rpm-ostree[2037]: client(id:cli dbus:1.55 unit:machine-config-daemon-firstboot.service uid:0) vanished; remaining=0
Jul 30 06:10:38 master1.ocp4.beer4laff.de rpm-ostree[2037]: In idle state; will auto-exit in 63 seconds
Jul 30 06:11:42 master1.ocp4.beer4laff.de rpm-ostree[2037]: In idle state; will auto-exit in 61 seconds
Jul 30 06:11:42 master1.ocp4.beer4laff.de systemd[1]: rpm-ostreed.service: Succeeded.
-- Subject: Unit succeeded
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
-- 
-- The unit rpm-ostreed.service has successfully entered the 'dead' state.
Jul 30 06:11:42 master1.ocp4.beer4laff.de systemd[1]: rpm-ostreed.service: Consumed 48ms CPU time
-- Subject: Resources consumed by unit runtime
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
-- 
-- The unit rpm-ostreed.service completed and consumed the indicated resources.
[core@master1 ~]$

**Comment by apoorvajagtap** [comment]
Hello, as rpm-ostreed service did not indicate any primary concerns, possible to share the sosreport of the affected master node?
That might help with a better analysis of the logs. As an option, can you try uploading the sosreport to google drive and sharing that here?

**Comment by olafWW** [comment]
I opend a RH case to further analyze this.. I'll provide the feedback here asap ... many many thx sofar ...

**Comment by fdammeke** [comment]
@olafWW Did you receive any feedback? I'm bumping into the exact same issue.

**Comment by olafWW** [comment]
Hi - @fdammeke - the issue is still open - after ton's of debug sessions with RHEl support .. its still open/ unsolved - I'll post any progress/solutions here asap

**Comment by olafWW** [comment]
@fdammeke  I made progress and was able to successfully install the cluster.. Obviously the segmetation fault is not the root cause - I changed 2 things
I enhanced the dhcpd.conf, that the hostname and the domain is also included (in addition to the  just fixed IP) 
smth like this:

host master1 {
  hardware ethernet 12:20:02:11:34:11;
  fixed-address 10.0.11.51;
  option domain-name "ocp4.beer4laff.de";
  option host-name "master1";
}

2ndly .. I created a new haproxy file from scratch (using the documented version on https://docs.openshift.com/container-platform/4.8/installing/installing_bare_metal/installing-bare-metal.html#cluster-entitlements_installing-bare-metal
